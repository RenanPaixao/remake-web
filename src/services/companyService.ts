import { supabase } from '../utils/supabase.ts'
import { Location, LocationsService } from './locationService.ts'

export interface Company {
  id: string
  name: string
  created_at?: string
  owner_id: string
}

export interface CompanyWithLocations extends Company {
  locations: Location[]
}

class PlacesServiceImp {
  get queryBuilder() {
    return supabase.from('companies')
  }
  async getAllWithLocations(options?: {from:number, to:number}): Promise<{ count: number, companies: CompanyWithLocations[] }> {
    const { data: companies, count } = options ?
      await this.queryBuilder.select('*, locations(*)', { count: 'exact' }).range(options.from, options.to).throwOnError() :
      await this.queryBuilder.select('*, locations(*)', { count: 'exact' }).throwOnError()

    if (companies === null) {
      return {
        companies: [],
        count: 0
      }
    }

    return {
      companies: companies.filter(company => company.locations.length),
      count: count ?? 0
    }
  }

  async searchCompanies(query: string): Promise<{ count: number, companies: CompanyWithLocations[] }> {
    const { data: companies, count } = await this.queryBuilder
      .select('*, locations(*)', { count: 'exact' })
      .textSearch('name', `${query}`, {
        type: 'websearch'
      })
      .throwOnError()

    if (companies === null) {
      return {
        companies: [],
        count: 0
      }
    }

    return {
      companies: companies.filter(company => company.locations.length),
      count: count ?? 0
    }
  }

  async deleteCompany(id: string): Promise<void> {
    await this.queryBuilder.delete().eq('id', id).throwOnError()
  }

  async createCompanyWithLocation(
    company: Omit<Company, 'id'>,
    location: Omit<Location, 'company_id'>): Promise<void> {
    const { data, error } = await this.queryBuilder.insert(company).select().maybeSingle()

    if(error) {
      throw new Error(error.message)
    }

    try {
      await LocationsService.create({
        ...location,
        company_id: data.id
      })
    }catch(e: any) {
      await this.deleteCompany(data.id)

      if(e.message) {
        throw new Error(e.message)
      }
      throw new Error('Error creating company')
    }
  }

  async getUserPlaces(userId: string): Promise<CompanyWithLocations[]> {
    const { data: companies } = await this.queryBuilder
      .select('*, locations(*)')
      .eq('owner_id', userId)
      .throwOnError()

    if (companies === null) {
      return []
    }

    return companies.filter(company => company.locations.length)
  }
}

export const CompanyService = new PlacesServiceImp()
