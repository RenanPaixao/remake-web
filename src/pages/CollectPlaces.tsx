import { Box, Center } from '@chakra-ui/react'
import { PlaceCard } from '../components/PlaceCard/PlaceCard.tsx'
import { useEffect, useState } from 'react'
import { CompanyService, CompanyWithLocations } from '../services/companyService.ts'
import haversine from 'haversine'
import { Pagination } from '../components/Pagination/Pagination.tsx'

export const CollectPlaces = () => {
  const [companies, setCompanies] = useState<CompanyWithLocations[]>([])
  const [currentPosition, setPosition] = useState({ latitude: 0, longitude: 0 })
  const [totalPages, setTotalPages] = useState(0)
  const [currentPage, setCurrentPage] = useState(1)

  useEffect(() => {
    (async() => {
      const data = await CompanyService.getAllWithLocations({ from: 0, to: 9 })
      setCompanies(data)

      try {
        const total = await CompanyService.countCompanies()
        setTotalPages(Math.ceil(total / 10))
      }catch(e: any) {
        console.error(e)
      }
    })()

    const location = sessionStorage.getItem('location')
    if(!location) {
      console.error('No location found')
      return
    }

    setPosition(JSON.parse(location))
  }, [])

  useEffect(() => {
    (async() => {
      const from = (currentPage - 1) * 10
      const to = from + 10
      const data = await CompanyService.getAllWithLocations({ from, to })
      setCompanies(data)
    })()
  }, [currentPage])

  /**
   * Get the distance between the current position and the given position.
   * @param position
   */
  function getDistance(position: { latitude: number, longitude: number}) {
    const distance = haversine(
      currentPosition,
      position
    )

    return `${distance.toFixed(2)} km`
  }

  /**
   * Go to the previous page.
   */
  async function onPrevious(): Promise<void> {
    setCurrentPage(prevState => prevState - 1)
  }

  /**
   * Go to the next page.
   */
  async function onNext(): Promise<void> {
    setCurrentPage(prevState => prevState + 1)
  }

  /**
   * Go to the given page.
   * @param number
   */
  async function onNumberClick(number: number): Promise<void> {
    if(number === currentPage) {
      return
    }

    setCurrentPage(number)
  }

  return <Center py={20} flexDirection={'column'}>
    {
      companies.map(company => {
        return company.locations.map(location => {
          const locationCoords = {
            latitude: location.latitude,
            longitude: location.longitude
          }

          return <Center key={location.id} w={'100%'} my={2}>
            <PlaceCard title={company.name} address={location.name} distance={getDistance(locationCoords)}/>
          </Center>
        })
      })
    }

    <Box mt={'2rem'}>
      <Pagination
        size={totalPages}
        activeNumber={currentPage}
        onPrevious={onPrevious}
        onNext={onNext}
        onNumberClick={onNumberClick}
        gap={3}
      />
    </Box>
  </Center>
}
