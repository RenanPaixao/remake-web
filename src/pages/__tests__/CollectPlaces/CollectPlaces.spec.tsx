import { CollectPlaces } from '../../CollectPlaces.tsx'
import { customRender } from '../../../../tests/test-utils.tsx'
import { COMPANIES, POSITION } from '../../../components/Pagination/__tests__/mock.ts'
import { CompanyService } from '../../../services/companyService.ts'
import { waitFor } from '@testing-library/dom'
import { expect } from 'vitest'

CompanyService.getAllWithLocations = vi.fn().mockImplementation(async () => {
  setTimeout(() => {
    return Promise.resolve({
      companies: COMPANIES,
      count: 100
    })
  }, 2000)
}).mockResolvedValue({
  companies: COMPANIES,
  count: 100
})
sessionStorage.getItem = vi.fn().mockReturnValue(JSON.stringify(POSITION))

describe('CollectPlaces', () => {
  it('should render', async () => {
    const { container, getByRole } = customRender(<CollectPlaces/>)

    const loading = getByRole('generic', { name: 'loading' })

    await waitFor(() => {
      expect(loading).not.toBeInTheDocument()
    })

    expect(container).toMatchSnapshot()
  })
})
