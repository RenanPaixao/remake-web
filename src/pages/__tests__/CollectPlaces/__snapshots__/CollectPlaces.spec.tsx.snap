// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CollectPlaces > should render 1`] = `
<div>
  <div
    class="css-1nfposj"
  >
    <div
      class="css-jecjnx"
    >
      <div
        class="chakra-input__group css-klzvhf"
        data-group="true"
      >
        <div
          class="chakra-input__left-element css-gv1iwz"
        >
          <svg
            color="var(--chakra-colors-gray-400)"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
        </div>
        <input
          aria-label="search bar"
          class="chakra-input css-tiw55w"
          placeholder="Search for a place"
          type="search"
          value=""
        />
        <div
          class="chakra-input__right-addon css-x8mdnz"
        >
          <button
            class="chakra-button css-jut409"
            type="button"
          >
            Search
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-tsn4jp"
    >
      <div
        class="chakra-card css-1101n89"
      >
        <div
          class="chakra-card__header css-1fj4epu"
        >
          <h3
            class="chakra-heading css-1q6o3tu"
          >
            testedoteste
          </h3>
          <h5
            class="chakra-text css-1vuos2j"
          >
            4003.62 km
          </h5>
        </div>
        <div
          class="chakra-card__footer css-siy0oy"
        >
          <p
            class="chakra-text css-cg4qa7"
          >
            rua
          </p>
          <button
            class="chakra-button css-1noe2cj"
            type="button"
          >
            Go!
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-tsn4jp"
    >
      <div
        class="chakra-card css-1101n89"
      >
        <div
          class="chakra-card__header css-1fj4epu"
        >
          <h3
            class="chakra-heading css-1q6o3tu"
          >
            Empresa - Derby
          </h3>
          <h5
            class="chakra-text css-1vuos2j"
          >
            3969.79 km
          </h5>
        </div>
        <div
          class="chakra-card__footer css-siy0oy"
        >
          <p
            class="chakra-text css-cg4qa7"
          >
            principal teste
          </p>
          <button
            class="chakra-button css-1noe2cj"
            type="button"
          >
            Go!
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-tsn4jp"
    >
      <div
        class="chakra-card css-1101n89"
      >
        <div
          class="chakra-card__header css-1fj4epu"
        >
          <h3
            class="chakra-heading css-1q6o3tu"
          >
            Empresa Renan
          </h3>
          <h5
            class="chakra-text css-1vuos2j"
          >
            10686.77 km
          </h5>
        </div>
        <div
          class="chakra-card__footer css-siy0oy"
        >
          <p
            class="chakra-text css-cg4qa7"
          >
            fake street
          </p>
          <button
            class="chakra-button css-1noe2cj"
            type="button"
          >
            Go!
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-tsn4jp"
    >
      <div
        class="chakra-card css-1101n89"
      >
        <div
          class="chakra-card__header css-1fj4epu"
        >
          <h3
            class="chakra-heading css-1q6o3tu"
          >
            Empresa Renan
          </h3>
          <h5
            class="chakra-text css-1vuos2j"
          >
            7525.17 km
          </h5>
        </div>
        <div
          class="chakra-card__footer css-siy0oy"
        >
          <p
            class="chakra-text css-cg4qa7"
          >
            boca do acre
          </p>
          <button
            class="chakra-button css-1noe2cj"
            type="button"
          >
            Go!
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-tsn4jp"
    >
      <div
        class="chakra-card css-1101n89"
      >
        <div
          class="chakra-card__header css-1fj4epu"
        >
          <h3
            class="chakra-heading css-1q6o3tu"
          >
            nova empresa
          </h3>
          <h5
            class="chakra-text css-1vuos2j"
          >
            3996.40 km
          </h5>
        </div>
        <div
          class="chakra-card__footer css-siy0oy"
        >
          <p
            class="chakra-text css-cg4qa7"
          >
            Rua de Paudalho
          </p>
          <button
            class="chakra-button css-1noe2cj"
            type="button"
          >
            Go!
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-tsn4jp"
    >
      <div
        class="chakra-card css-1101n89"
      >
        <div
          class="chakra-card__header css-1fj4epu"
        >
          <h3
            class="chakra-heading css-1q6o3tu"
          >
            Casa amarela
          </h3>
          <h5
            class="chakra-text css-1vuos2j"
          >
            3971.59 km
          </h5>
        </div>
        <div
          class="chakra-card__footer css-siy0oy"
        >
          <p
            class="chakra-text css-cg4qa7"
          >
            Rua Guaimbé
          </p>
          <button
            class="chakra-button css-1noe2cj"
            type="button"
          >
            Go!
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-tsn4jp"
    >
      <div
        class="chakra-card css-1101n89"
      >
        <div
          class="chakra-card__header css-1fj4epu"
        >
          <h3
            class="chakra-heading css-1q6o3tu"
          >
            Absgc
          </h3>
          <h5
            class="chakra-text css-1vuos2j"
          >
            3973.50 km
          </h5>
        </div>
        <div
          class="chakra-card__footer css-siy0oy"
        >
          <p
            class="chakra-text css-cg4qa7"
          >
            Avenida Governador Agamenon Magalhães
          </p>
          <button
            class="chakra-button css-1noe2cj"
            type="button"
          >
            Go!
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-tsn4jp"
    >
      <div
        class="chakra-card css-1101n89"
      >
        <div
          class="chakra-card__header css-1fj4epu"
        >
          <h3
            class="chakra-heading css-1q6o3tu"
          >
            Ahah
          </h3>
          <h5
            class="chakra-text css-1vuos2j"
          >
            157.25 km
          </h5>
        </div>
        <div
          class="chakra-card__footer css-siy0oy"
        >
          <p
            class="chakra-text css-cg4qa7"
          >
            Rua Antônio Carneiro Cézar Menezes
          </p>
          <button
            class="chakra-button css-1noe2cj"
            type="button"
          >
            Go!
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-tsn4jp"
    >
      <div
        class="chakra-card css-1101n89"
      >
        <div
          class="chakra-card__header css-1fj4epu"
        >
          <h3
            class="chakra-heading css-1q6o3tu"
          >
            teste
          </h3>
          <h5
            class="chakra-text css-1vuos2j"
          >
            3525.28 km
          </h5>
        </div>
        <div
          class="chakra-card__footer css-siy0oy"
        >
          <p
            class="chakra-text css-cg4qa7"
          >
            Rua Caxambu
          </p>
          <button
            class="chakra-button css-1noe2cj"
            type="button"
          >
            Go!
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-tsn4jp"
    >
      <div
        class="chakra-card css-1101n89"
      >
        <div
          class="chakra-card__header css-1fj4epu"
        >
          <h3
            class="chakra-heading css-1q6o3tu"
          >
            joao
          </h3>
          <h5
            class="chakra-text css-1vuos2j"
          >
            3959.14 km
          </h5>
        </div>
        <div
          class="chakra-card__footer css-siy0oy"
        >
          <p
            class="chakra-text css-cg4qa7"
          >
            Rua Caxambu
          </p>
          <button
            class="chakra-button css-1noe2cj"
            type="button"
          >
            Go!
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-tsn4jp"
    >
      <div
        class="chakra-card css-1101n89"
      >
        <div
          class="chakra-card__header css-1fj4epu"
        >
          <h3
            class="chakra-heading css-1q6o3tu"
          >
            Lucas
          </h3>
          <h5
            class="chakra-text css-1vuos2j"
          >
            4217.20 km
          </h5>
        </div>
        <div
          class="chakra-card__footer css-siy0oy"
        >
          <p
            class="chakra-text css-cg4qa7"
          >
            Rua Japaratuba
          </p>
          <button
            class="chakra-button css-1noe2cj"
            type="button"
          >
            Go!
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-i51og3"
    >
      <div
        class="css-1fpbagy"
      >
        <button
          aria-label="previous button"
          class="chakra-button css-19byqlw"
          disabled=""
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
            />
          </svg>
        </button>
        <button
          aria-label="button page 1"
          class="chakra-button css-19byqlw"
          data-active=""
          type="button"
        >
          1
        </button>
        <button
          aria-label="button page 2"
          class="chakra-button css-19byqlw"
          type="button"
        >
          2
        </button>
        <button
          aria-label="button page 3"
          class="chakra-button css-19byqlw"
          type="button"
        >
          3
        </button>
        <button
          aria-label="button page 4"
          class="chakra-button css-19byqlw"
          type="button"
        >
          4
        </button>
        <button
          aria-label="button page 5"
          class="chakra-button css-19byqlw"
          type="button"
        >
          5
        </button>
        <button
          aria-label="button page 6"
          class="chakra-button css-19byqlw"
          type="button"
        >
          6
        </button>
        <button
          aria-label="button page 7"
          class="chakra-button css-19byqlw"
          type="button"
        >
          7
        </button>
        <button
          aria-label="button page 8"
          class="chakra-button css-19byqlw"
          type="button"
        >
          8
        </button>
        <button
          aria-label="button page 9"
          class="chakra-button css-19byqlw"
          type="button"
        >
          9
        </button>
        <button
          aria-label="button page 10"
          class="chakra-button css-19byqlw"
          type="button"
        >
          10
        </button>
        <button
          aria-label="next button"
          class="chakra-button css-19byqlw"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
